plugins/modules/win_domain_computer.ps1

```diff
@@ -146,6 +156,82 @@ Function Add-ConstructedState($desired_state) {
   $result.changed = $true
 }
 
+Function Invoke-OfflineDomainJoin {
+  [CmdletBinding(SupportsShouldProcess=$true)]
+  param(
+    [Parameter(Mandatory=$true)]
+    [System.Collections.IDictionary]
+    $desired_state ,
+
+    [Parameter(Mandatory=$true)]
+    [ValidateSet('none','output','path')]
+    [String]
+    $Action ,
+
+    [Parameter()]
+    [System.IO.FileInfo]
+    $BlobPath
+  )
+
+  End {
+    if ($Action -eq 'none') {
+      return
+    }
+
+    $dns_domain = $desired_state.dns_hostname -replace '^[^.]+\.'
+
+    $output = $Action -eq 'output'
+
+    $arguments = @(
+      'djoin.exe'
+      '/PROVISION'
+      '/REUSE'  # we're pre-creating the machine normally to set other fields, then overwriting it with this
+      '/DOMAIN'
+      $dns_domain
+      '/MACHINE'
+      $desired_state.sam_account_name.TrimEnd('$')  # this machine name is the short name
+      '/MACHINEOU'
+      $desired_state.ou
+      '/SAVEFILE'
+      $BlobPath.FullName
+    )
+
+    $invocation = Argv-ToString -arguments $arguments
+    $result.djoin = @{
+      invocation = $invocation
+    }
+    $result.odj_blob = ''
+
+    if ($Action -eq 'path') {
+      $result.odj_blob_path = $BlobPath.FullName
+    }
+
+    if ($PSCmdlet.ShouldProcess($argstring)) {
+      try {
+        $djoin_result = Run-Command -command $invocation
+        $result.djoin.rc = $djoin_result.rc
+        $result.djoin.stdout = $djoin_result.stdout
+        $result.djoin.stderr = $djoin_result.stderr
+
+        if ($djoin_result.rc) {
+          Fail-Json -obj $result -message "Problem running djoin.exe. See returned values."
+        }
+
+        if ($output) {
+          $bytes = [System.IO.File]::ReadAllBytes($BlobPath.FullName)
+          $data = [Convert]::ToBase64String($bytes)
+          $result.odj_blob = $data
+        }
+      }
+      finally {
+        if ($output) {
+          $BlobPath.Delete()
```

Comment:
It seems like this file may not actually exist in a failure scenario. When looking at the code we are passing through a path which could be defined by the end user. If `djoin.exe` hasn't created the file for some reason (say it failed) then this will fail as well hiding the actual error.

Also, I'm unsure whether we need to check if the parent directory exists and fail if it doesn't. Does `djoin.exe` create any intermediate parent dirs that don't exist?

Reply:
Good call; `djoin` will not create intermediate paths. I've added a check for that.

One of the things I like about the `FileInfo.Delete()` method is that it doesn't fail if the file doesn't exist; but to your point it will fail if the parent directory doesn't exist; so added a check for file existence anyway.

Also updated the docs to indicate the parent of the BLOB path needs to exist.